### ISSUE TITLE
If pending connection appears, all others metrics are set to 0

### ISSUE THREAD

**Issue by jeremie-H** [brettwooldridge/HikariCP#897]
In production mode, our hikari pool is under heavy load, and when the maximum seems to be reached, metrics returned by dropwizard are weird. Only 1 pending connection, and all others metrics are set to 0.
is the problem with dropwizard or HikariCP ?

#### Environment
```
HikariCP version: 2.4.3
JDK version     : 1.8.0_65
Database        : Oracle 12.1.0.2.0
```
-----------------------------------------------------------------------------------------

Normal case :
datasourcePRODUCTION.pool.ActiveConnections: {
value: 1
},
datasourcePRODUCTION.pool.IdleConnections: {
value: 10
},
datasourcePRODUCTION.pool.PendingConnections: {
value: 0
},
datasourcePRODUCTION.pool.TotalConnections: {
value: 10
},

Pending case : 
datasourcePRODUCTION.pool.ActiveConnections: {
value: 0
},
datasourcePRODUCTION.pool.IdleConnections: {
value: **0**
},
datasourcePRODUCTION.pool.PendingConnections: {
value: 1
},
datasourcePRODUCTION.pool.TotalConnections: {
value: **0**
},

**Comment by brettwooldridge** [comment]
Can you please try on the latest library version?

**Comment by AlexanderMann** [comment]
Not sure if this is the same, but I'm getting a similar behaviour.

DB: Postgresql AWS RDS
Env: `clojure:lein-2.7.1` -> `openjdk:8`
HikariCP Version: `[hikari-cp "1.7.5"]` `[com.zaxxer/HikariCP "2.5.1"]`
Dropwizard Version: `[io.dropwizard.metrics/metrics-core "3.2.3"]`
PSQL JDBC Version: `[org.postgresql/postgresql "9.4.1212"]`

I have a reported spitting stats every 15 seconds, but the values I get for my active connections are virtually zilch. However, I know that the number of requests I'm serving (successfully) is in the thousands per minute...so I assume that it's actually talking to the db (more so since I have timing and call stats for the our db roundtrips...)

Two separate services running in production show the following graphs.:

<img width="1259" alt="screen shot 2017-08-04 at 15 41 27" src="https://user-images.githubusercontent.com/3885029/28986261-ceb737a0-792b-11e7-8d7a-7383d59fe840.png">
<img width="1266" alt="screen shot 2017-08-04 at 15 44 05" src="https://user-images.githubusercontent.com/3885029/28986262-ced180e2-792b-11e7-8663-f524d5d65cdd.png">

**Comment by AlexanderMann** [comment]
Going to try and bump all the things and see if that fixes.

**Comment by AlexanderMann** [comment]
Confirmed that HikariCP version is: `[com.zaxxer/HikariCP "2.6.3"]`

Still seeing:

<img width="1261" alt="screen shot 2017-08-04 at 16 56 01" src="https://user-images.githubusercontent.com/3885029/28988614-3819c672-7936-11e7-921f-cb3f90954293.png">
<img width="1265" alt="screen shot 2017-08-04 at 16 56 43" src="https://user-images.githubusercontent.com/3885029/28988615-381a64ec-7936-11e7-98fc-e6e45957cfb6.png">

**Comment by AlexanderMann** [comment]
Hey @brettwooldridge,

Still working on getting you the logs etc., but thought you might find this interesting/helpful. Last night we had a production "blip" where Usage and Wait times skyrocketed and we saw our pods get rotated due to:

```
java.sql.SQLTransientConnectionException: HikariPool-1 - Connection is not available, request timed out after 30000ms.
```

We had 55 instances of that error over about a 30 second period. This occurred over 5 of our 7 pods in loosely the same distribution (ranging from 7 at min to 17 at max). Behaviour during this was that our Wait and Usage times skyrocketed (eventually coming back down. More interestingly, the TimeoutRate has stayed solid after this at 23. Not entirely sure why, or what's up, etc.

Pool's config is:

```
  {:auto-commit        true
   :read-only          false
   :connection-timeout 30000
   :validation-timeout 5000
   :idle-timeout       600000
   :max-lifetime       1800000
   :minimum-idle       10
   :maximum-pool-size  10
   :register-mbeans    false}
```

Now, all of this matters because, during this period I was able to get this distribution graph from my connections:

<img width="1263" alt="screen shot 2017-08-05 at 09 02 25" src="https://user-images.githubusercontent.com/3885029/28995999-d5d63c98-79bc-11e7-882b-9892eb25a870.png">

I have tests setup to conduct the logging you were after, so please stay tuned! I'm wondering now if my problem is just that I'm just missing the counters when they have values I'm after etc. Maybe a historgram metric is more what I was expecting?

**Comment by brettwooldridge** [comment]
Here's a little insight into how the statistics are collected.  Whenever you call dropwizard to collect stats, [this method](https://github.com/brettwooldridge/HikariCP/blob/dev/src/main/java/com/zaxxer/hikari/metrics/PoolStats.java#L82) is called. The logic is, if it has been more than one second since the last call, refresh the stats, otherwise return the previous stats. So, basically stats are refreshed on demand at a maximum rate of once per second.

**Comment by brettwooldridge** [comment]
Re: above, the stats are a snapshot of the pool state at the instant that the metrics are requested, and as far as I can tell are as accurate as possible given the dynamic nature of the pool.  But note, there is no synchronization, so it is possible for active+idle to be > total.

**Comment by AlexanderMann** [comment]
@brettwooldridge yeah, I think the above is accurate. I pulled the logs you were asking for but nothing funky stood out.

I think what I'm probably after (as far as "active" and "pending" counts) is probably _best_ gotten by the Usage.Count and Wait.Count metrics. Any other recommendations I'm happy to hear them, but please feel free to close this on my part.

Also, goes without saying, but Hikari is awesome, and I really appreciate you as the maintainer being so responsive and reasonably opinionated on the operation of db pooling etc.

**Comment by pratyushbansal** [comment]
> Hey @brettwooldridge,
> 
> Still working on getting you the logs etc., but thought you might find this interesting/helpful. Last night we had a production "blip" where Usage and Wait times skyrocketed and we saw our pods get rotated due to:
> 
> ```
> java.sql.SQLTransientConnectionException: HikariPool-1 - Connection is not available, request timed out after 30000ms.
> ```
> 
> We had 55 instances of that error over about a 30 second period. This occurred over 5 of our 7 pods in loosely the same distribution (ranging from 7 at min to 17 at max). Behaviour during this was that our Wait and Usage times skyrocketed (eventually coming back down. More interestingly, the TimeoutRate has stayed solid after this at 23. Not entirely sure why, or what's up, etc.
> 
> Pool's config is:
> 
> ```
>   {:auto-commit        true
>    :read-only          false
>    :connection-timeout 30000
>    :validation-timeout 5000
>    :idle-timeout       600000
>    :max-lifetime       1800000
>    :minimum-idle       10
>    :maximum-pool-size  10
>    :register-mbeans    false}
> ```
> 
> Now, all of this matters because, during this period I was able to get this distribution graph from my connections:
> 
> <img alt="screen shot 2017-08-05 at 09 02 25" width="1263" src="https://user-images.githubusercontent.com/3885029/28995999-d5d63c98-79bc-11e7-882b-9892eb25a870.png">
> 
> I have tests setup to conduct the logging you were after, so please stay tuned! I'm wondering now if my problem is just that I'm just missing the counters when they have values I'm after etc. Maybe a historgram metric is more what I was expecting?

@AlexanderMann I know I am really late, but if possible could you please share the root cause analysis of your problem. Thanks

**Comment by AlexanderMann** [comment]
Oof, @pratyushbansal this was quite a while ago and at an employer past...best I can _remember_ we switched things over to not use the active vs pending connections. The better metric was what @brettwooldridge recommended of using the:

> Usage.Count and Wait.Count

I _think_...it's been so long now though.

The incident above where I got the weirdness in production is also something I can't remember.
